### ISSUE TITLE
Standalone API deployment tested locally with MiniKube

### ISSUE THREAD

**Issue by gctucker** [kernelci/kernelci-api#274]
This is the API counterpart of #273, to replicate the standard `docker-compose` deployment in a local MiniKube Kubernetes cluster.  If #273 is done first, then the pipeline can be running in MiniKube too and connect to the API services within the same cluster.  In principle this is likely to be the preferred way of deploying the API and pipeline services from a performance point of view.  Having them in separate clusters might help reduce downtime, for example if we only need to migrate the pipeline.  These are things we can already start considering and experiment with using MiniKube before recreating it in the Cloud with Azure as a follow-up task.

- [x] K8s manifest files for API: https://github.com/kernelci/kernelci-api/pull/287
- [x] Tested with `kci` command line on the host, connecting to the API service inside MiniKube e.g. `kci api hello`
- [x] Tested with full pipeline services running also in MiniKube if possible, and/or in `docker-compose` on the host
- [x] Configuration file created for automatically deploying the API services added to `kernelci-api` repository
- [x] Summary wrt deploying both pipeline and API services in Kubernetes

**Comment by gctucker** [comment]
I've added a note to say that the LAVA callback handler and the KCIDB bridge services aren't part of this task as they're special cases and they're not part of the main pipeline.

Once this task is done, a possible follow-up would be this one to add the LAVA callback handler: https://github.com/kernelci/kernelci-api/issues/282

In a way it's the simplest service we need to deploy which needs to accept incoming HTTPS connections, so it might be good preparation work for then adding the API services (main API has HTTPS, then there's SSH and another HTTP for storage with static files).

**Comment by gctucker** [comment]
(comment above was meant to be here: https://github.com/kernelci/kernelci-api/issues/273#issuecomment-1603901746)

**Comment by JenySadadia** [comment]
Used `kompose` tool to create services and deployment definition files for all the containers.
```
$ kompose convert -f ../docker-compose.yaml --volumes=hostPath
WARN Service "redis" won't be created because 'ports' is not specified 
INFO Kubernetes file "api-service.yaml" created   
INFO Kubernetes file "db-service.yaml" created    
INFO Kubernetes file "ssh-service.yaml" created   
INFO Kubernetes file "storage-service.yaml" created 
INFO Kubernetes file "api-deployment.yaml" created 
INFO Kubernetes file "env-configmap.yaml" created 
INFO Kubernetes file "db-deployment.yaml" created 
INFO Kubernetes file "redis-deployment.yaml" created 
INFO Kubernetes file "ssh-deployment.yaml" created 
INFO Kubernetes file "storage-deployment.yaml" created 

```
Tested `api` container with Minikube.
```
$ kubectl apply -f api-service.yaml,api-deployment.yaml 
service/api created
deployment.apps/api created
```
```
$ kubectl describe svc api
Name:              api
Namespace:         default
Labels:            io.kompose.service=api
Annotations:       kompose.cmd: kompose convert -f docker-compose.yaml
                   kompose.version: 1.26.0 (40646f47)
Selector:          io.kompose.service=api
Type:              ClusterIP
IP Family Policy:  SingleStack
IP Families:       IPv4
IP:                10.108.172.216
IPs:               10.108.172.216
Port:              8001  8001/TCP
TargetPort:        8000/TCP
Endpoints:         <none>
Session Affinity:  None
Events:            <none>
```
```
$ curl http://localhost:8001/latest/
{"message":"KernelCI API"}
```

**Comment by gctucker** [comment]
As a side note, from the description:

> experiment with using MiniKube before recreating it in the Cloud with Azure as a follow-up task

I think we're still missing that follow-up task to actually deploy things in Azure and sort out SSL for public incoming connections etc.  I tried to create an ingress definition via the Azure portal but I didn't appear to have permissions for setting up an SSL certificate for some obscure reasons.

**Comment by gctucker** [comment]
@JenySadadia @sbdtu5498 Hello, could you please go through the checklist in the issue definition and confirm which ones have already been done?

**Comment by JenySadadia** [comment]
> @JenySadadia @sbdtu5498 Hello, could you please go through the checklist in the issue definition and confirm which ones have already been done?

I have updated the checklist. I am testing the full pipeline atm. 
@sbdtu5498  For `Configuration file created for automatically deploying the API services added to kernelci-api repository`,
maybe we can add `.properties` file to generate configmap variables from it?
A shell script can also be added to execute `kubectl` commands to generate secrets and configmap.

**Comment by sbdtu5498** [comment]
@JenySadadia can you elaborate the point a bit more.

**Comment by sbdtu5498** [comment]
> @JenySadadia @sbdtu5498 Hello, could you please go through the checklist in the issue definition and confirm which ones have already been done?

First two are done, I will be reordering the pipeline on Wednesday or before the next community huddle. Pipeline manifests are already there, but I need to take a glance, fix all nits, reorder the directory structure, add README.md and scripts as well. After that minikube deployment task would be completely done for the pipeline and API. So by this week I hope that Pipeline stuff would be merged as well.

**Comment by JenySadadia** [comment]
> > @sbdtu5498 For Configuration file created for automatically deploying the API services added to kernelci-api repository,
> > maybe we can add .properties file to generate configmap variables from it?
> > A shell script can also be added to execute kubectl commands to generate secrets and configmap.
> 
> @JenySadadia can you elaborate the point a bit more. The current apply-all.sh can get the kernelci-api automatically up and running. If we add an env file that can be passed to the shell or maybe a configmap, we can also incorporate other env variables.

Please check this link https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/
We can add a YAML file with `kind: ConfigMap` and add `REDIS_HOST` and `MONGO_SERVICE` to it and read that file to generate configmap.

**Comment by sbdtu5498** [comment]
> > > @sbdtu5498 For Configuration file created for automatically deploying the API services added to kernelci-api repository,
> > > maybe we can add .properties file to generate configmap variables from it?
> > > A shell script can also be added to execute kubectl commands to generate secrets and configmap.
> > 
> > 
> > @JenySadadia can you elaborate the point a bit more. The current apply-all.sh can get the kernelci-api automatically up and running. If we add an env file that can be passed to the shell or maybe a configmap, we can also incorporate other env variables.
> 
> Please check this link https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/ We can add a YAML file with `kind: ConfigMap` and add `REDIS_HOST` and `MONGO_SERVICE` to it and read that file to generate configmap.

Got your point. There are two ways to do it. One is to use Kustomize, or one is to simply use configmap. I will add whatever seems better. Actually both seems equivalently easy to be used by the end user or us.

**Comment by JenySadadia** [comment]
> > > > @sbdtu5498 For Configuration file created for automatically deploying the API services added to kernelci-api repository,
> > > > maybe we can add .properties file to generate configmap variables from it?
> > > > A shell script can also be added to execute kubectl commands to generate secrets and configmap.
> > > 
> > > 
> > > @JenySadadia can you elaborate the point a bit more. The current apply-all.sh can get the kernelci-api automatically up and running. If we add an env file that can be passed to the shell or maybe a configmap, we can also incorporate other env variables.
> > 
> > 
> > Please check this link https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/ We can add a YAML file with `kind: ConfigMap` and add `REDIS_HOST` and `MONGO_SERVICE` to it and read that file to generate configmap.
> 
> Got your point. There are two ways to do it. One is to use Kustomize, or one is to simply use configmap. I will add whatever seems better. Actually both seems equivalently easy to be used by the end user or us.

The below file would work in case of pipeline:
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: kernelci-pipeline-config
data:
  settings: "/home/kernelci/config/kernelci.toml"

```

**Comment by sbdtu5498** [comment]
> apiVersion: v1
kind: ConfigMap
metadata:
  name: kernelci-pipeline-config
data:
  settings: "/home/kernelci/config/kernelci.toml"
  
Yes this would be a perfect approach as it takes all the env in the existing way.

**Comment by gctucker** [comment]
> `settings: "/home/kernelci/config/kernelci.toml"`

It's also possible to put the file in a standard location e.g. `/home/kernelci/.config/kernelci/kernelci.toml` and not have to specify it.

**Comment by sbdtu5498** [comment]
> > apiVersion: v1
> > kind: ConfigMap
> > metadata:
> > name: kernelci-pipeline-config
> > data:
> > settings: "/home/kernelci/config/kernelci.toml"
> 
> Yes this would be a perfect approach as it takes all the env in the existing way.

I just realized that it would be better to simply use SETTINGS as an Env like

env:
- name: SETTINGS
- value: "<required location>"

It would have same impact as creating a one liner config map

**Comment by gctucker** [comment]
Ah yes, if the env variable is then passed as command line arguments to the various pipeline services.  If there are blocking issues around this we could just rely on the default paths to get started and then add configuration as a follow-up.

**Comment by gctucker** [comment]
@JenySadadia Could you please update the checklist?  I believe it's all done in which case we should be able to close this issue.

**Comment by JenySadadia** [comment]
Yes, it's all done. Updated the checklist.
Closing the issue.
